robot/impl/robot_reconfigure_weak_dependencies_test.go

```diff
@@ -285,4 +291,363 @@ func TestUpdateWeakDependents(t *testing.T) {
+	// 3) weak1 will configure.
+	// 4) At the end of reconfiguration, weak1 will reconfigure once.
+	t.Log("Robot startup")
+	base1Name := base.Named("base1")
+	base2Name := base.Named("base2")
+	weakCfg1 := config.Config{
+		Components: []resource.Config{
+			{
+				Name:                weak1Name.Name,
+				API:                 weakAPI,
+				Model:               weakModel,
+				DependsOn:           []string{base1Name.String()},
+				ConvertedAttributes: &someTypeWithWeakAndStrongDepsConfig{},
+			},
+			{
+				Name:  base1Name.Name,
+				API:   base.API,
+				Model: fake.Model,
+			},
+			{
+				Name:  base2Name.Name,
+				API:   base.API,
+				Model: fake.Model,
+			},
+		},
+	}
+	test.That(t, weakCfg1.Ensure(false, logger), test.ShouldBeNil)
+	robot := setupLocalRobot(t, context.Background(), &weakCfg1, logger)
+
+	res, err := robot.ResourceByName(weak1Name)
+	// The resource was found and all dependencies were properly resolved.
+	test.That(t, err, test.ShouldBeNil)
+	weak1, err := resource.AsType[*someTypeWithWeakAndStrongDeps](res)
+	test.That(t, err, test.ShouldBeNil)
+	// Assert that the weak dependency was tracked.
+	test.That(t, weak1.resources, test.ShouldHaveLength, 2)
+	test.That(t, weak1.resources, test.ShouldContainKey, base1Name)
+	test.That(t, weak1.reconfigCount, test.ShouldEqual, 1)
+
+	// Reconfigure base1
+	// The following scenario is expected:
+	// 1) The base1 will reconfigure (which will bump the logical clock).
+	// 2) The completeConfig loop will check to see if there is a need to updateWeakDependents by
+	//    checking the logical clock on also whether there are dependencies on weak dependents.
+	//    The check will fail.
+	// 3) weak1 will not reconfigure, as base1 was not newly built or error during reconfiguration.
+	// 4) At the end of reconfiguration, weak1 will reconfigure once.
+	t.Log("Reconfigure base1")
+	weakCfg2 := config.Config{
+		Components: []resource.Config{
+			{
+				Name:                weak1Name.Name,
+				API:                 weakAPI,
+				Model:               weakModel,
+				DependsOn:           []string{base1Name.String()},
+				ConvertedAttributes: &someTypeWithWeakAndStrongDepsConfig{},
+			},
+			{
+				Name:  base1Name.Name,
+				API:   base.API,
+				Model: fake.Model,
+				// We need the following `robot.Reconfigure` to call `Reconfigure` on this `base1`
+				// component. We change the `Attributes` field from the previous (nil) value to
+				// accomplish that.
+				Attributes: rutils.AttributeMap{"version": 1},
+			},
+			{
+				Name:  base2Name.Name,
+				API:   base.API,
+				Model: fake.Model,
+			},
+		},
+	}
+	test.That(t, weakCfg2.Ensure(false, logger), test.ShouldBeNil)
+	robot.Reconfigure(context.Background(), &weakCfg2)
+
+	res, err = robot.ResourceByName(weak1Name)
+	test.That(t, err, test.ShouldBeNil)
+	weak1, err = resource.AsType[*someTypeWithWeakAndStrongDeps](res)
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, weak1.reconfigCount, test.ShouldEqual, 2)
```

Comment:
Ok, so this has now advanced to 2. Which I assume is because its weak dependency changed. So I'm confused what the story for this step of the test is communicating.

Reply:
we run updateWeakDependents at the end of reconfiguration, which is why this has been bumped. let me know if the comments communicate the story better now.